design/resource-filter-field-selectors-design.md

```diff
@@ -0,0 +1,229 @@
+# Extend Include/Exclude Filters With Field Selectors
+
+## Abstract
+
+This design proposal extends the cluster-scoped and namespace-scoped resource filters described in [#5333][0], to support resource filtering via [field selectors][3].
+It provides users with more granular control over resources to be included in and excluded from backup and restore operations.
+
+## Background
+
+The efficiency and reliability of backup and restore operations constitute a critical part of all data protection pipelines on large production Kubernetes clusters.
+To satisfy the SLOs of these ever-growing sophisticated backup and restore scenarios, Velero should provide the users with the ability to control the size and scope of these operations, with more granular selection of the targeted resources.
+
+The existing [label selection][1] mechanism supported by Velero doesn't scale on large production Kubernetes clusters, due to constraints beyond the users' control like inconsistent resource labeling scheme across different providers, strict cluster-scoped RBAC permissions, dynamically generated custom resource definitions etc.
+
+## Goals
+
+- Users can define filter criteria like resource name and namespace to include and exclude resources from backup and restore, using Kubernetes field selectors.
+
+## Non Goals
+
+- N/A
+
+## High-Level Design
+
+This design proposes changing the `[]string` type of the following fields in the `BackupSpec` API to a structure that can encapsulate the field selector properties:
```

Comment:
What about putting all the filed selector properties into [ResourcePolicies](https://github.com/vmware-tanzu/velero/blob/main/design/Implemented/handle-backup-of-volumes-by-resources-filters.md?plain=1#L69),
and all these field selector properties form a new kind of resource policies. 
For example:
```golang
type ResourcePolicies struct {
        Version        string         `yaml:"version"`
	VolumePolicies []VolumePolicy `yaml:"volumePolicies"`
         FieldSelectorPolicies []FieldSelectorPolicy `yaml:"fieldSelectorPolicies"`
	// we may support other resource policies in the future, and they could be added separately
  	// OtherResourcePolicies: []OtherResourcePolicy
}
````

As more and more filters are added now, we tend to consolidate all future filters into the ResourcePolicies struct. This will make the configuration more centralized and avoid the backup spec from storing a large number of filters, which can cause its size to increase.

Reply:
When using resource policy, is it possible for us to not define the intermediate YAML file, instead just create the config map directly?

Reply:
@ihcsim Yes, the YAML file is a template for users to define policy and then generate a config map from the YAML file.
Also, users could create a config map directly, but the data in the config map should be parsed by Velero.

Reply:
@ihcsim 

Would you refine the design to apply the "resource-policy" approach?

As for whether or not it will be in v1.12, we can make decisions based on the milestone we defined in the wiki.

Reply:
I will take a closer look at the resource policy. Also, does this limitation (mentioned in the 1.11 release changelog) has something to do with the creation of the configmap?

> The Flexible resource policy that can filter volumes to skip in the backup is not guaranteed to work on unofficial third-party plugins because the plugins may not follow the existing backup workflow code logic of Velero. The ConfigMap used as the policy is supposed to be maintained by users.

Reply:
@ihcsim It is because some unofficial third-party plugins may not implement Velero standard backup and restore interface in the plugin. Velero will filter resources by resource policy when calling the standard b&r interface. They may have their own CRDs to do b&r (such as the vSphere plugin), so the standard filter resources may not work. 

Our officially supported plugins ([AWS](https://github.com/vmware-tanzu/velero-plugin-for-aws), [Azure](https://github.com/vmware-tanzu/velero-plugin-for-microsoft-azure), [GCP](https://github.com/vmware-tanzu/velero-plugin-for-gcp), [vSphere](https://github.com/vmware-tanzu/velero-plugin-for-vsphere)) are functioning well.